# 1/6/22
# Purpose: make files for batch submission for downloading, aligning, and peak calling

# cd /home/ak2267/project/Roller/ChIP/batch
# python prepFiles_processRoller.py E-MTAB-7127.sdrf.txt

import sys

inFileSummary = open(sys.argv[1],'rt') # opens E-MTAB-7127.sdrf.txt to get file info

Sample_Dict = {}
# Sample_Dict[species][tissue][mark][replicate] = fastq

for line in inFileSummary:
    if line[0] == 'd':
        splitLine = line.strip().split('\t')
        species = str(splitLine[3])
        tissue = str(splitLine[10].split(" ")[0])
        if splitLine[41][0] == 'i':
            mark = 'input'
        else:
            mark = str(splitLine[41].split('-')[1])
        replicate = str(splitLine[11])
        fastq = str(splitLine[30])

        if species not in Sample_Dict:
            Sample_Dict[species] = {}
        if tissue not in Sample_Dict[species]:
            Sample_Dict[species][tissue] = {}
        if mark not in Sample_Dict[species][tissue]:
            Sample_Dict[species][tissue][mark] = {}
        if replicate not in Sample_Dict[species][tissue][mark]:
            Sample_Dict[species][tissue][mark][replicate] = fastq

# Translate full species names to UCSC abbreviation
Species_Translator = {'Callithrix jacchus':'calJac4','Canis lupus familiaris':'canFam6','Equus caballus':'equCab3','Felis catus':'felCat9','Macaca mulatta':'rheMac10','Monodelphis domestica':'monDom5','Mus musculus':'mm39','Oryctolagus cuniculus':'oryCun2','Rattus norvegicus':'rn7','Sus scrofa':'susScr11'}

# Effective genome size generated by unique-kmers.py
Effective_GenomeSize = {'calJac4':2621714877,'canFam6':2240314204,'equCab3':2365156725,'felCat9':2352800598,'rheMac10':2653677440,'mm39':2309746861,'rn7':2369256514,'susScr11':2359453421}

outDownload = open('220712_downloadFastq_jobFile.txt','wt')
outFastqc = open('220712_fastqc_jobFile.txt','wt')
outAlign = open('220712_align_jobFile.txt','wt')
outSort = open('220713_sort_jobFile.txt','wt')
outFilter = open('220713_filter_jobFile.txt','wt')
outCallPeaks = open('220713_callPeaks_jobFile.txt','wt')
outVisualize = open('220714_visualize_jobFile.txt','wt')

for species in Sample_Dict:
    if species != 'Monodelphis domestica' and species != 'Oryctolagus cuniculus':
        for tissue in Sample_Dict[species]:
            for mark in Sample_Dict[species][tissue]:
                for replicate in Sample_Dict[species][tissue][mark]:

                    fileString = str(Species_Translator[species])+'_'+str(tissue)+'_'+str(mark)+'_'+str(replicate)
                    #print(fileString)

                    # make DOWNLOAD job file for use with deadSimpleQueue (dSQ)
                    outDownload.write("cd /gpfs/gibbs/pi/noonan/ak2267/Roller/fastq ; source /home/ak2267/.bash_profile ; source /home/ak2267/.bashrc ; wget "+str(Sample_Dict[species][tissue][mark][replicate])+" ; mv "+str(Sample_Dict[species][tissue][mark][replicate].split('/')[-1])+" "+fileString+".fastq.gz"+'\n')

                    # make FASTQC job file
                    outFastqc.write("cd /gpfs/gibbs/pi/noonan/ak2267/Roller/fastq ; source /home/ak2267/.bash_profile ; source /home/ak2267/.bashrc ; module load FastQC/0.11.9-Java-11 ; fastqc "+fileString+".fastq.gz -o fastqc"+'\n')

                    # make ALIGN job file - remove '--no-unal' and change to '--very-sensitive'
                    outAlign.write("cd /gpfs/gibbs/pi/noonan/ak2267/Roller/bam ; source /home/ak2267/.bash_profile ; source /home/ak2267/.bashrc ; module load Bowtie2/2.4.2-GCCcore-10.2.0 ; module load SAMtools/1.13-GCCcore-10.2.0 ; bowtie2 --very-sensitive -p 8 -x /gpfs/gibbs/pi/noonan/ak2267/genomes/"+str(Species_Translator[species])+" -U /gpfs/gibbs/pi/noonan/ak2267/Roller/fastq/"+fileString+".fastq.gz 2> log/"+fileString+"_bowtie2.log | samtools view -@7 -b > "+fileString+"_bowtie2.bam"+'\n')

                    # make SORT job file to sort bams (then remove bam files to free space)
                    outSort.write("cd /gpfs/gibbs/pi/noonan/ak2267/Roller/bam ; source /home/ak2267/.bash_profile ; source /home/ak2267/.bashrc ; module load SAMtools/1.13-GCCcore-10.2.0 ; module load Sambamba/0.7.1 ; sambamba sort -t 2 -o "+fileString+"_bowtie2_sorted.bam "+fileString+"_bowtie2.bam"+'\n')

                    # make FILTER job file - 1st step = remove multimappers and unmapped; 2nd step = remove duplicates
                    outFilter.write("cd /gpfs/gibbs/pi/noonan/ak2267/Roller/bam ; source /home/ak2267/.bash_profile ; source /home/ak2267/.bashrc ; module load SAMtools/1.13-GCCcore-10.2.0 ; module load Sambamba/0.7.1 ; sambamba view -h -t 2 -f bam -F \"[XS] == null and not unmapped\" "+fileString+"_bowtie2_sorted.bam > "+fileString+"_bowtie2_intermediateFiltered.bam 2> log/"+fileString+"_intermediateFiltered.log ; sambamba markdup -r -t 2 "+fileString+"_bowtie2_intermediateFiltered.bam "+fileString+"_bowtie2_filtered.bam 2> log/"+fileString+"_filtered.log "+'\n')

                    # make CALL PEAKS job file - remove '--bdg' for now to save space, modify to use filtered bam files
                    if mark != 'input':
                        # broad for H3K4me1
                        if mark == 'H3K4me1':
                            outCallPeaks.write("cd /gpfs/gibbs/pi/noonan/ak2267/Roller/peaks ; source /home/ak2267/.bash_profile ; source /home/ak2267/.bashrc ; module load MACS2/2.2.7.1-foss-2020b-Python-3.8.6 ; module load SAMtools/1.13-GCCcore-10.2.0 ; macs2 callpeak -t /gpfs/gibbs/pi/noonan/ak2267/Roller/bam/"+fileString+"_bowtie2_filtered.bam -c /gpfs/gibbs/pi/noonan/ak2267/Roller/bam/"+str(Species_Translator[species])+'_'+str(tissue)+'_input_'+str(replicate)+"_bowtie2_filtered.bam -n "+fileString+" --broad -g "+str(Effective_GenomeSize[Species_Translator[species]])+' 2> log/'+fileString+'.log'+'\n')
                        # narrow for H3K4me3 and H3K27ac - just remove "--broad", there's no "--narrow" flag
                        if mark == 'H3K4me3' or mark == 'H3K27ac':
                            outCallPeaks.write("cd /gpfs/gibbs/pi/noonan/ak2267/Roller/peaks ; source /home/ak2267/.bash_profile ; source /home/ak2267/.bashrc ; module load MACS2/2.2.7.1-foss-2020b-Python-3.8.6 ; module load SAMtools/1.13-GCCcore-10.2.0 ; macs2 callpeak -t /gpfs/gibbs/pi/noonan/ak2267/Roller/bam/"+fileString+"_bowtie2_filtered.bam -c /gpfs/gibbs/pi/noonan/ak2267/Roller/bam/"+str(Species_Translator[species])+'_'+str(tissue)+'_input_'+str(replicate)+"_bowtie2_filtered.bam -n "+fileString+" -g "+str(Effective_GenomeSize[Species_Translator[species]])+' 2> log/'+fileString+'.log'+'\n')
                    
                    # make VISUALIZE job file - makes bigWigs and bigBeds for use on the genome browser
                    # added --extendReads and --centerReads, removed --smoothLength 60
                    # had to edit this for input files that came as a and b so I only make 1 bigWig
                    if mark != 'input':
                        peakType = {'H3K4me3':'narrow','H3K27ac':'narrow','H3K4me1':'broad'}
                        outVisualize.write("cd /gpfs/gibbs/pi/noonan/ak2267/Roller/visualize ; source /home/ak2267/.bash_profile ; source /home/ak2267/.bashrc ; module load deepTools/3.5.1-foss-2020b ; bamCoverage -b /gpfs/gibbs/pi/noonan/ak2267/Roller/bam/"+fileString+"_bowtie2_filtered.bam -o "+"/gpfs/gibbs/pi/noonan/ak2267/Roller/visualize/"+fileString+".bw --effectiveGenomeSize "+str(Effective_GenomeSize[Species_Translator[species]])+" --normalizeUsing CPM --binSize 10 --extendReads 300 --centerReads -p 2 2> log/"+fileString+"_bamCoverage.log ; cut -f 1,2,3,4 /gpfs/gibbs/pi/noonan/ak2267/Roller/peaks/"+fileString+"_peaks."+peakType[mark]+"Peak > temp_"+fileString+".bed ; bedToBigBed temp_"+fileString+".bed /home/ak2267/genomes/chrom.sizes/"+str(Species_Translator[species])+".chrom.sizes "+fileString+"_"+peakType[mark]+"Peak.bb"+'\n')
                    elif mark == 'input' and fileString[-1] != 'b':
                        if fileString[-1] == 'a':
                            fileString = fileString[:-1]
                        outVisualize.write("cd /gpfs/gibbs/pi/noonan/ak2267/Roller/visualize ; source /home/ak2267/.bash_profile ; source /home/ak2267/.bashrc ; module load deepTools/3.5.1-foss-2020b ; bamCoverage -b /gpfs/gibbs/pi/noonan/ak2267/Roller/bam/"+fileString+"_bowtie2_filtered.bam -o "+"/gpfs/gibbs/pi/noonan/ak2267/Roller/visualize/"+fileString+".bw --effectiveGenomeSize "+str(Effective_GenomeSize[Species_Translator[species]])+" --normalizeUsing CPM --binSize 10 --extendReads 300 --centerReads -p 2 2> log/"+fileString+"_bamCoverage.log"+'\n')

outDownload.close()
outFastqc.close()
outAlign.close()
outSort.close()
outFilter.close()
outCallPeaks.close()
outVisualize.close()

